circleci.go

```diff
@@ -0,0 +1,319 @@
+		err = db.UpsertArtifacts(artifacts)
+		if err != nil {
+			return nil, err
+		}
+
+		err = db.ArchiveBuild(b.BuildNum)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return artifacts, nil
+}
+
+func (c *Client) BootstrapBuilds(db RootDB) error {
+	return db.RunInTx(func(tx DB) error {
+		println("bootstrapping database, this may take a while...")
+		var nbuilds int
+		var nreported int
+
+		// Our goal is linking to artifacts, which last for 30 days.
+		boundary := time.Now().AddDate(0, 0, -30)
+
+		for offset := 0; true; offset += 100 {
+			var builds []Build
+			builds, err := c.GetBuilds(100, offset)
+			if err != nil {
+				return err
+			}
+			if len(builds) == 0 {
+				break
+			}
+			for _, b := range builds {
+				if !b.StartTime.After(boundary) {
+					fmt.Printf(
+						"Quitting after 30 days of builds; artifacts won't exist further back.\n",
+					)
+					goto report
+				}
+				err = tx.UpsertBuild(b)
+				if err != nil {
+					return err
+				}
+				nbuilds++
+			}
+			if nbuilds >= nreported + 1000 {
+				nreported = nbuilds
+				fmt.Printf("%v builds found so far...\n", nbuilds)
+			}
+		}
+
+	report:
+		fmt.Printf("bootstrapping database complete, %v builds found.\n", nbuilds)
+		return nil
+	})
+}
+
+func (c *Client) RefreshBuilds(db RootDB, afterBuildNum int) (error) {
+	return db.RunInTx(func(tx DB) error {
+		for offset := 0; true; offset += 100 {
+			builds, err := c.GetBuilds(100, offset)
+			if err != nil {
+				return err
+			}
+			if len(builds) == 0 {
+				return errors.New("ran out of builds during a refresh??")
+			}
+			for _, b := range builds {
+				if b.BuildNum <= afterBuildNum {
+					return nil
+				}
+				err = tx.UpsertBuild(b)
+				if err != nil {
+					return err
+				}
+			}
+		}
+		return nil
+	})
+}
+
+/*
+Strategy:
+- on boostrap, scrape as far back as you can go for coverage artifacts
+- periodically scrape as far back as the oldest non-completed build in the database
+- have a force-refresh page that triggers the scrape when a user loads the page
+*/
+
+func (c *Client) StartupDB(dbPath string) (RootDB, error) {
+	db, err := NewDB(dbPath)
+	if err != nil {
+		return db, err
+	}
+
+	err = c.RefreshDB(db, true)
+	if err != nil {
+		return db, err
+	}
+
+	return db, nil
+}
+
+func (c *Client) RefreshDB(db RootDB, allowBootstrap bool) error {
+	oldestUnfinished, err := db.OldestUnfinishedBuild()
+	if err != nil {
+		return err
+	}
+
+	latestFinished, err := db.LatestFinishedBefore(oldestUnfinished)
+	if err != nil {
+		return err
+	}
+
+	if latestFinished == nil {
+		// nothing in the database, time for a bootstrap
+		if !allowBootstrap {
+			return errors.New("database still needs bootstrap!")
+		}
+		err = c.BootstrapBuilds(db)
+		if err != nil {
+			return err
+		}
+	} else {
+		err = c.RefreshBuilds(db, latestFinished.BuildNum)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (c *Client) RefreshBuildsThread(db RootDB, cond *sync.Cond, ready *bool) {
+	setReady := func() {
```

Comment:
No, it could not, because it's not a pipeline at all.

The artifact caching caches the most recent cacheable item it can at all times.  If it runs out of stuff to cache, it pauses until the next c.GetBuilds() kicks it (where it checks again and usually pauses again).

Effectively this strategy favors the newest artifacts during a the initial artifact bootstrap, which can take tens of minutes.  Assuming newer PRs are more important than 29-day-old PRs, this means the server is usable a bit sooner.

Reply:
talked in zoom.  Channels support a "try-send" behavior, so this could be done on a channel with buffer of size 1.